src/desktopentry.cpp

```diff
@@ -0,0 +1,193 @@
+    auto it = std::find(components.cbegin(), components.cend(), "applications");
+    if (it == components.cend()) {
+        qWarning() << "fatal error: file location invalid";
+        return ParseError::InvalidLocation;
+    }
+    QString FileId;
+    ++it;
+    while (it != components.cend())
+        FileId += (*(it++) + "-");
+    m_desktopId = FileId.chopped(1);
+
+    auto file = QFile{m_filePath};
+    if (!file.open(QFile::ReadOnly | QFile::Text)) {
+        qWarning() << "failed to open desktop file:" << file.errorString();
+        return ParseError::OpenFailed;
+    }
+
+    QTextStream in{&file};
+    in.setEncoding(QStringConverter::Utf8);
+    decltype(entryMap)::iterator currentGroup;
+
+    DesktopEntry::ParseError err{ParseError::NoError};
+    while (!in.atEnd()) {
+        auto line = in.readLine().trimmed();
+
+        if (line.startsWith("[")) {
+            if (!line.endsWith("]"))
+                return ParseError::GroupHeaderInvalid;
+            currentGroup = parserGroupHeader(line);
+        } else {
+            if (auto e = parseEntry(line, currentGroup); e != ParseError::NoError) {
+                err = e;
+                qWarning() << "an error occurred,this line will be skipped:"<< line;
+            }
+        }
+    }
+    return err;
+}
+
+QMap<QString, DesktopEntry::Value> DesktopEntry::group(const QString &key) const noexcept
+{
+    if (auto group = entryMap.find(key); group != entryMap.cend())
+        return *group;
+    return {};
+}
+
+QString DesktopEntry::Value::unescape(const QString& str) const noexcept
+{
+    constexpr static char escapeChar[]{'s','n','t','r','\\',';'};
+    QString unescapedStr;
+    for (std::size_t i = 0; i < str.size(); ++i) {
+        auto c = str.at(i);
+        if (c == '\\') {
+            auto e = std::find(std::begin(escapeChar), std::end(escapeChar), str.at(i + 1));
+            if(e != std::end(escapeChar)){
+                switch (*e) {
+                    case 'n':
+                        unescapedStr.append('\n');
+                        break;
+                    case 't':
+                        unescapedStr.append('\t');
+                        break;
+                    case 'r':
+                        unescapedStr.append('\r');
+                        break;
+                    case '\\':
+                        unescapedStr.append('\\');
+                        break;
+                    case ';':
+                        unescapedStr.append(';');
+                        break;
+                    case 's':
+                        unescapedStr.append(' ');
+                        break;
+                }
+                ++i;
+            } else {
+                unescapedStr.append(c);
+            }
+        }else {
+            unescapedStr.append(c);
+        }
+    }
+
+    return unescapedStr;
+}
+
+QString DesktopEntry::Value::toString(const QString& key) const noexcept
+{
```

Comment:
I think the `key` is not that useful in this method.

When user try to parse a localestring as string, I think this toString method should just return the one without any \[].

And a user should not get the localed value of this key.

Reply:
this key has a default value